src/Text/Layout/Table/Cell/WideString.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiWayIf                 #-}
+{-# LANGUAGE OverloadedStrings          #-}
+module Text.Layout.Table.Cell.WideString
+    ( WideString(..)
+    , WideText(..)
+    ) where
+
+import Data.String
+import qualified Data.Text as T
+import Text.DocLayout
+
+import Text.Layout.Table.Cell
+import Text.Layout.Table.Primitives.AlignInfo
+import Text.Layout.Table.StringBuilder
+
+-- | A newtype for String in which characters can be wider than one space.
+newtype WideString = WideString String
+    deriving (Eq, Ord, Show, Read, Semigroup, Monoid, IsString, StringBuilder)
+
+instance Cell WideString where
+    dropLeft i (WideString s) = WideString $ dropWide True i s
+    dropRight i (WideString s) = WideString . reverse . dropWide False i $ reverse s
+    visibleLength (WideString s) = realLength s
+    measureAlignment p (WideString s) = measureAlignmentWide p s
+
+    buildCell (WideString s) = buildCell s
+
+dropWide :: Bool -> Int -> String -> String
+dropWide _ i [] = []
+dropWide gobbleZeroWidth i (x:xs)
+    | gobbleZeroWidth && i == 0 && l == 0 = dropWide gobbleZeroWidth i xs
+    | i <= 0    = x:xs
+    | l <= i    = dropWide gobbleZeroWidth (i - l) xs
+    | otherwise = replicate (l - i) ' ' ++ xs
+  where l = charWidth x
+
+measureAlignmentWide :: (Char -> Bool) -> String -> AlignInfo
+measureAlignmentWide p xs = case break p xs of
+    (ls, rs) -> AlignInfo (realLength ls) $ case rs of
+        []      -> Nothing
+        _ : rs' -> Just $ realLength rs'
+
+-- | A newtype for Text in which characters can be wider than one space.
+newtype WideText = WideText T.Text
+    deriving (Eq, Ord, Show, Read, Semigroup, Monoid, IsString, StringBuilder)
+
+instance Cell WideText where
+    dropLeft i (WideText s) = WideText $ dropLeftWideT i s
+    dropRight i (WideText s) = WideText $ dropRightWideT i s
+    visibleLength (WideText s) = realLength s
+    measureAlignment p (WideText s) = measureAlignmentWideT p s
+
+    buildCell (WideText s) = buildCell s
+
+dropLeftWideT :: Int -> T.Text -> T.Text
+dropLeftWideT i txt = case T.uncons txt of
+    Nothing -> txt
+    Just (x,xs) -> let l = charWidth x in if
+        | i == 0 && l == 0 -> dropLeftWideT i xs
+        | i <= 0    -> txt
+        | l <= i    -> dropLeftWideT (i - l) xs
+        | otherwise -> T.replicate (l - i) " " <> xs
+
+dropRightWideT :: Int -> T.Text -> T.Text
+dropRightWideT i txt = case T.unsnoc txt of
+    Nothing -> txt
+    Just (xs,x) -> let l = charWidth x in if
+        | i <= 0    -> txt
+        | l <= i    -> dropRightWideT (i - l) xs
+        | otherwise -> xs <> T.replicate (l - i) " "
```

Comment:
These are identical except for the `otherwise` case.  Can you think of some way to refactor that?

Reply:
I'll try to think of a way. They're actually more different than they look: note that one uses `uncons` and the other uses `unsnoc`. This means their pattern matches are reversed too (one is `(x,xs)`, the other is `(xs,x)`).

Reply:
@Xitian9, oh, my bad.  Then just forget about it. :)